### ISSUE TITLE
Request of EXIF support for convert JXL to JPG (exif needs to be added after convertion)

### ISSUE THREAD

**Issue by MartinEesmaa** [libjxl/libjxl#540]
Hello! I'm looking forward to builders needs to add DJXL for exif support after convertion from JXL to JPG, so JPG recompressed need to have same original metadata came from my camera picture. I tried using exiftool to add .exif metadata into JXL file, but updated file failed.

3 Major Developers, that I need request help to add EXIF support from JXL to JPG
@deymo 
@jonsneyers 
@jxl-bot 

Please add support to build.

Regards
- Martin Eesmaa

**Comment by nathanielcwm** [comment]
what?

**Comment by jloqfjgk** [comment]
@jxl-bot is a bot, so there is no need to ping it.

-----

If you would like to manually edit EXIF metadata, libjxl is not the right tool. Use exiftool, exiv2, etc. instead.

Since JXL is a new format, many tools do not support it well. For example [exiftool](https://github.com/exiftool/exiftool) can only read but not write metadata for JXL, and [exiv2](https://github.com/Exiv2/exiv2) only supports JXL/bmff ([JXL/codestream is a TODO](https://github.com/Exiv2/exiv2/issues/1506)).

-----

On the other hand, if you take photos with your camera and it contains EXIF metadata, when you `cjxl image.jpg image.jxl` (lossless JPEG transcode) or `djxl image.jxl image.jpg` (get back the original JPG), all the metadata is kept.

**Comment by MartinEesmaa** [comment]
Impossible, I tried djxl image.jxl image.jpg of 8000x6000, but it successed convertion, but metadata failed.
Result: JPEG XL decoder v0.6.0 6e1e309 [Scalar]
Read 3871181 compressed bytes.
Warning: could not decode losslessly to JPEG. Retrying with --pixels_to_jpeg...
Decoded to pixels.
8000 x 6000, 79.68 MP/s [79.68, 79.68], 1 reps, 8 threads.
Allocations: 533 (max bytes in use: 6.993952E+08)

Exiftool image.jpg converted:
ExifTool Version Number         : 12.30
File Name                       : image.jpg
Directory                       : .
File Size                       : 11 MiB
File Modification Date/Time     : 2021:09:04 22:54:56+10:00
File Access Date/Time           : 2021:09:04 22:55:16+10:00
File Creation Date/Time         : 2021:09:04 22:54:56+10:00
File Permissions                : -rw-rw-rw-
File Type                       : JPEG
File Type Extension             : jpg
MIME Type                       : image/jpeg
JFIF Version                    : 1.01
Resolution Unit                 : None
X Resolution                    : 1
Y Resolution                    : 1
Warning                         : Malformed APP1 EXIF segment
Image Width                     : 8000
Image Height                    : 6000
Encoding Process                : Baseline DCT, Huffman coding
Bits Per Sample                 : 8
Color Components                : 3
Y Cb Cr Sub Sampling            : YCbCr4:4:4 (1 1)
Image Size                      : 8000x6000
Megapixels                      : 48.0

**Comment by TheHardew** [comment]
@jloqfjgk exif is not kept if you transcode to jxl in a lossy way and decode that to another jpeg.

@MartinEesmaa I forgot to tell you that decoding to jpeg after a lossy transcodeing to jxl will further decrease the quality, so just in general avoid doing that.

**Comment by thebombzen** [comment]
@MartinEesmaa The design is that `cjxl image.jpg image.jxl` should be fully reversible with `djxl image.jxl image.jpg`. That is, the file produced from a lossless jpeg transcode and back should be bit-exact identical to the original. If you have a sample that makes this not true, then please provide it as an attachment.

**Comment by MartinEesmaa** [comment]
Yeah, only default settings works that I tried cjxl to JXL file and DJXL to JPG file, the metadata kept still after convert from JXL to JPG, but the default settings of JXL gave me: 8.13 MB, so I want make smallest file of JXL and best quality, so I posted thread a couple days ago: [](https://www.reddit.com/r/jpegxl/comments/pgvrpu/what_is_the_best_jxl_compression_settings_without/)

I know the resampling of JXL file (652 KB), that browser decodes 1/4 quarter photo.  What is best settings to encode JXL for smallest file without resampling? Also it needs convert JXL small file to JPG for closed metadata?

**Comment by jloqfjgk** [comment]
See also: https://old.reddit.com/r/jpegxl/comments/pgvrpu/what_is_the_best_jxl_compression_settings_without/

**Comment by MartinEesmaa** [comment]
I already posted my thread.

**Comment by jloqfjgk** [comment]
@MartinEesmaa, are you describing the following issue?

* PNG with EXIF and/or XMP -> `cjxl` to JXL (metadata is kept) -> djxl to new PNG (metadata is kept)
* JPG with EXIF -> `cjxl -d 1 image.jpg image.jxl` (metadata is kept) -> djxl to new PNG / JPG (metadata is **not** kept)

**Comment by MartinEesmaa** [comment]
I just encoded to JXL by following code: `cjxl -d 3 --resampling=2 image.jpg image.jxl` and `djxl image.jxl image.png` PNG only works that metadata kept, but JPG is not kept metadata after convert from JXL to JPG.

**Comment by thebombzen** [comment]
A lossleess JPG transcode has only one quality, which is lossless. If you specify the distance yourself (via `-d 3`) then you're not doing a lossless transcode anymore, you're performing a new encoding. You won't be able to reconstruct the original jpeg perfectly, nor would you want to, seeing as that would throw away a bunch of quality. After calling `cjxl -d 3 image.jpg image.jxl` you should either keep it as JXL, or decode directly to PNG.

**Comment by MartinEesmaa** [comment]
Yeah after PNG, how can I compress PNG to JPG to kept metadata still after convert PNG to JPG?

**Comment by thebombzen** [comment]
Why would you want to do that? As in, why are you converting to JXL at all if your're just going to (lossily) transcode to JPG?

**Comment by MartinEesmaa** [comment]
I'm not sure why I wanted to JXL always, example my January 2021 photos size is 24 GB, I want to save up to 10 GB using encode to JXL, oh by the way reason if I convert back JXL to PNG, PNG takes me 32 MB than original 10 MB JPG/652 KB JXL.

**Comment by thebombzen** [comment]
If you're encoding to JXL lossily, you should just leave it in the JXL format. Decoding it after the fact to PNG and then encoding it to JPG doesn't accomplish anything, when you can just do JPG -> PNG -> JPG and skip a step of generation loss.

**Comment by MartinEesmaa** [comment]
Okay, umm I have to edit sample photo using Adobe Lightroom and Topaz Denoise AI and cjxl default settings. Is it correct?

**Comment by megapro17** [comment]
Also there's a problem when cjxl doesn't read exif from png files

**Comment by jloqfjgk** [comment]
@jonsneyers said on JXL Discord:
> For Exif, there's another issue: if you have e.g. a 90 degree oriented JPEG and losslessly transcode it to jxl, then the jxl will have the orientation in the jxl codestream header, but also in the Exif. When decoding to pixels, the Exif is ignored and we use the codestream header to properly orient the image (this allows you to just strip all Exif if you want, without getting a wrongly oriented image).
When djxl writes a lossy transcoded jpeg, it first decodes to pixels, and then encodes those pixels in jpeg
So it does orient the pixels and produces a jpeg that is already oriented (not like the original jpeg that needed exif to be oriented)
That's fine, but if we would now just add the Exif to that jpeg, then viewers would orient the jpeg again, causing it to be rotated 180 degrees instead of 90
One possible solution could be to not orient decoded pixels if we know we're going to write a jpeg and we're going to add Exif
Another solution would be to change the Exif we write to do no orientation.

> To be honest, I don't really know why djxl supports writing a lossy jpeg at all. It's not really its job to be a jpeg encoder.

> It would be convenient if djxl would let you extract metadata though
Like imagemagick does, you can do `convert foo.jpg foo.exif` to get the embedded exif blob
We could make `djxl` do that too quite easily

**Comment by TheHardew** [comment]
I'm a little surprised, that `djxl` correctly adds exif to png files then. As far as I can tell, png does not store any orientation data in its header, so the image is rotated before being compressed.

**Comment by jonsneyers** [comment]
It's complicated when making tooling that converts between formats:

- jxl is clear: specs says codestream orientation counts, Exif orientation shall be ignored
- jpeg: can contain Exif, most viewers (in particular all browsers) apply the Exif orientation
- png: can contain Exif, most viewers (in particular all browsers) ignore the Exif orientation

We were not doing an optimal job in doing sensible things in all cases, I hope after #544 it will be better.

**Comment by MartinEesmaa** [comment]
Okay, maybe I will wait for build to fix EXIF, so thank you, @jonsneyers! I definitely wait to get version 1 of CJXL and DJXL, so it should be better and the browsers may have to fix `--resampling d2` to see full image.

**Comment by jonsneyers** [comment]
Things should be better now that #544 is merged.
